internal/venice-common/src/main/resources/avro/KafkaMessageEnvelope/v12/KafkaMessageEnvelope.avsc

```diff
@@ -0,0 +1,371 @@
+                        "null",
+                        {
+                          "type": "array",
+                          "items": "bytes"
+                        }
+                      ],
+                      "default": null
+                    }, {
+                      "name": "isRepush",
+                      "doc": "Flag to indicate this version swap is triggered by repush or not.",
+                      "type": "boolean",
+                      "default": false
+                    }, {
+                      "name": "isLastVersionSwapMessageFromRealTimeTopic",
+                      "doc": "Flag to indicate this version swap message in version topic is triggered by the last version swap in real time topic the leader server has received. With this flag, new leader will be able to recover the full state during leadership handover, when we rely on real-time topics for all regions to achieve version swap synchronization.",
+                      "type": "boolean",
+                      "default": false
+                    }
+                  ]
+                }
+              ]
+            }
+          ]
+        }
+      ]
+    }, {
+      "name": "leaderMetadataFooter",
+      "doc": "A optional footer that leader SN can use to give extra L/F related mete data",
+      "type": [
+        "null",
+        {
+          "name": "LeaderMetadata",
+          "type": "record",
+          "fields": [
+            {
+              "name": "hostName",
+              "doc": "The identifier of the host which sends the message.This helps detect the 'split brain' scenario in leader SN. Notice that it is different from GUID. GUID represents the one who produces the message. In 'pass-through' mode, the relaying producer will reuse the same GUID from the upstream message.",
+              "type": "string"
+            }, {
+              "name": "upstreamOffset",
+              "doc": "Where this message is located in RT/GF/remote VT topic. This value will be determined and modified by leader SN at runtime.",
+              "type": "long",
+              "default": -1
+            }, {
+              "name": "upstreamKafkaClusterId",
+              "doc": "Kafka bootstrap server URL of the cluster where RT/GF/remote VT topic exists, represented by an integer to reduce the overhead. This value will be determined and modified by leader SN at runtime.",
+              "type": "int",
+              "default": -1
+            }, {
+              "name": "upstreamPubSubPosition",
+              "doc": "The position of the message in the upstream pubsub system (usually real-time topic).",
+              "type": ["null", "bytes"],
+              "default": null
+            }, {
+              "name": "termId",
+              "doc": "TermId is a unique identifier (usually a Helix message timestamp) for the term in which the message is produced.",
+              "type": "long",
+              "default": -1
+            }
```

Comment:
What is this `termId` for? Is it related to pub sub positions, or some separate thing? Is it related to improving leadership handovers?

Reply:
Yeah, it's is a separate thing intended to improve the leadership handover. Though this is not funded project, I thought since we're doing proto upgrade for KME it might be good idea to start adopting it as we are also going to do dataset refreshes later.

Reply:
More context around `termId`:

> `termId` is intended to uniquely represent a leader session or term in the replication log. In scenarios like **split-brain**, where multiple writers may exist simultaneously, `termId` will help establish **a partial order (happens-before relationship)** on records. Each replica will maintain a termId and update it upon becoming the leader, marking the start of a new leadership term. The leader will then attach this termId to every record it writes to the version topic, allowing followers to distinguish between records from the legitimate leader and those from outdated or conflicting writers. This mechanism is expected to ensure that replicas correctly identify and process only valid records.